internal/docker/client.go

```diff
@@ -0,0 +1,230 @@
+func Port(p string) ConfigOption {
+	return func(c *Config) (err error) {
+		c.Port, err = nat.NewPort("tcp", p)
+		return
+	}
+}
+
+// Env sets the environment variables to pass to the container.
+//
+// Example:
+//   - mysql: Env("MYSQL_ROOT_PASSWORD=password")
+//   - pg:    Env("POSTGRES_PASSWORD=password")
+func Env(env ...string) ConfigOption {
+	return func(c *Config) error {
+		c.Env = env
+		return nil
+	}
+}
+
+// Up pulls and starts a new docker container from the given image tag.
+func (c *Client) Up(ctx context.Context, cfg *Config) (*Container, error) {
+	if err := cfg.validate(); err != nil {
+		return nil, err
+	}
+	// Make sure the image is available
+	rd, err := c.c.ImagePull(ctx, cfg.Image, types.ImagePullOptions{})
+	_, err = io.Copy(os.Stdout, rd)
+	if err != nil {
+		return nil, err
+	}
+	// Get a free host TCP port the container can bind its exposed service port on.
+	p, err := freePort()
+	if err != nil {
+		return nil, err
+	}
+	// Create the container.
+	con, err := c.c.ContainerCreate(
+		ctx,
+		&container.Config{
+			Image: cfg.Image,
+			Env:   cfg.Env,
+		},
+		&container.HostConfig{
+			PortBindings: nat.PortMap{
+				cfg.Port: {{HostIP: "localhost", HostPort: p}},
+			},
+		},
+		nil, nil, "",
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := c.c.ContainerStart(ctx, con.ID, types.ContainerStartOptions{}); err != nil {
+		return nil, err
+	}
+	return &Container{
+		c:        c,
+		cfg:      *cfg,
+		ID:       con.ID,
+		Password: defaultPass,
+		Port:     p,
+	}, nil
+}
+
+// Down stops and removes a running container.
+func (c *Client) Down(ctx context.Context, ct *Container) error {
+	// Don't care about the error. ContainerRemove will get rid of the container anyway.
+	_ = c.c.ContainerStop(ctx, ct.ID, nil)
+	return c.c.ContainerRemove(ctx, ct.ID, types.ContainerRemoveOptions{})
+}
+
+// Down stops and removes this container.
+func (c *Container) Down(ctx context.Context) error {
+	return c.c.Down(ctx, c)
+}
+
+// Wait waits for this container to be ready.
+func (c *Container) Wait(ctx context.Context, timeout time.Duration) error {
+	mysqld.SetLogger(log.New(ioutil.Discard, "", 1))
+	defer mysqld.SetLogger(log.New(os.Stderr, "[mysql] ", log.Ldate|log.Ltime|log.Lshortfile))
+	if timeout > time.Minute {
+		timeout = time.Minute
+	}
+	done := time.After(timeout)
+	for {
+		select {
+		case <-time.After(100 * time.Millisecond):
+			db, err := sql.Open(c.Driver(), c.DSN())
+			if err != nil {
+				continue
+			}
+			if err := db.PingContext(ctx); err != nil {
+				continue
+			}
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-done:
+			return errors.New("timeout")
+		}
+	}
+}
+
+// Driver returns the database/sql driver name.
+func (c *Container) Driver() string {
+	return strings.SplitN(c.cfg.Image, ":", 2)[0]
+}
+
+// DSN returns a DSN to connect to the Container.
+func (c *Container) DSN() string {
+	return fmt.Sprintf("root:%s@tcp(localhost:%s)/", c.Password, c.Port)
+}
```

Comment:
Use `URL` instead. Also, this is a mysql driver format. Use the formats that exist here: https://atlasgo.io/cli/url

Reply:
It is not a URL in this case. It is meant to be used by `database/sql` which requires DSN, not URL. Atlas does.